### PR TITLE
Fix issue with cached partial collections

### PR DESCRIPTION
**Pull request by iainbryson**
It's possible to render a collection of partials with the `cached: true`
    parameter, each of which are also cached.  If those partials also
    have an `expires_in` value for their cache lifetime.  Because of the
    key collision between the partial's cache key and the individual
    collection elements' cache key, the non-expiring version will win.

    Fix is to allow cache parameters (well, only `expires_in`) to be passed
    allong with the cache collection render call. i.e., from:

    <%= render :partial => 'test/test', :collection => @models, :as => :test, :cached => true %>

    to:

    <%= render :partial => 'test/test2', :collection => @models, :as => :test, :cached => true, :expires_in => 5.seconds %>

    For #33424

    *Iain Bryson*

### Summary

Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together.

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails!

### CODE REVIEW

File: actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb [hunk 1]
```diff
@@ -14,6 +14,7 @@ module CollectionCaching # :nodoc:
       def cache_collection_render(instrumentation_payload)
         return yield unless @options[:cached]
 
+        @cache_options = @options.slice(:expires_in) # REVIEW: is Ruby 2.5-requiring code allowed here?
```
Comment (oniofchaos) [inline actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb]: Based on there being 2.4 builds on Travis CI, probably not :(

File: actionview/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,24 @@
...
+    collection elements' cache key, the non-expiring version will win.
+
+    Fix is to allow cache parameters (well, only `expires_in`) to be passed
+    allong with the cache collection render call. i.e., from:
+
+    <%= render :partial => 'test/test', :collection => @models, :as => :test, :cached => true %>
+
+    to:
+
+    <%= render :partial => 'test/test2', :collection => @models, :as => :test, :cached => true, :expires_in => 5.seconds %>
```
Comment (serene) [inline actionview/CHANGELOG.md]: How do you feel about changing this to:

`
render :partial => 'test/test2', :collection => @models, :as => :test, :cached => { :expires_in => 5.seconds }
`

**Comment by escanxr** [comment]
Does this PR still in work? I'm looking for a way to invalidate cached collection on specific actions (like if the locale change).

**Comment by serene** [comment]
@escanxr Maybe we should pull the changes into a new PR that is up to date with master?

**Comment by escanxr** [comment]
> @escanxr Maybe we should pull the changes into a new PR that is up to date with master?

I think so. I hope I will be available to do it this week-end

**Comment by jvillarejo** [comment]
Sorry for the up, but it would be good to have this. 
I have a collection that has a file attachment which I want to expire_in with the same `expires_in` configured for `ActiveStorage` 

Is there anyway I can help ?

**Comment by mpkhr** [comment]
+1

Ideally the expiration can be individual for each entry in the collection. Like the cache key name, which can be customized by giving a block to the render statement.

**Comment by jclusso** [comment]
This would be great to have, but as a work around, you can use something like this to force expire your cache every so often.

This would expire the cache every 5 minutes.

```ruby
= render collection: @posts, partial: 'post',
  cached: -> (post) { [post, (Time.now.min / 5)] }
```

**Comment by mpkhr** [comment]
Yep, that's what I currently do. Thanks for mentioning it though, so people know it's possible.

The request is in favor of recyclable cache keys over key-based expiration, which was a significant improvement in caching efficiency (DHH said: We went from only being able to keep 18 hours of caching to, I believe, 3 weeks. It was the single biggest performance boost that Basecamp 3 has ever seen.). So I think it's worth extending the collection-renderer with this capability

**Comment by jclusso** [comment]
I opened a new PR to do this. https://github.com/rails/rails/pull/51579

**Pull request closed without merging**